persistence/sql/process_local_queue.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2023 XDBLab organization
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sql
+
+import (
+	"context"
+	"github.com/xdblab/xdb-apis/goapi/xdbapi"
+	"github.com/xdblab/xdb/common/log/tag"
+	"github.com/xdblab/xdb/extensions"
+	"github.com/xdblab/xdb/persistence"
+)
+
+func (p sqlProcessStoreImpl) ProcessLocalQueueMessages(
+	ctx context.Context, request persistence.ProcessLocalQueueMessagesRequest,
+) (*persistence.ProcessLocalQueueMessagesResponse, error) {
+	tx, err := p.session.StartTransaction(ctx, defaultTxOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := p.doProcessLocalQueueMessageTx(ctx, tx, request)
+	if err != nil {
+		err2 := tx.Rollback()
+		if err2 != nil {
+			p.logger.Error("error on rollback transaction", tag.Error(err2))
+		}
+	} else {
+		err = tx.Commit()
+		if err != nil {
+			p.logger.Error("error on committing transaction", tag.Error(err))
+			return nil, err
+		}
+	}
+	return resp, err
+}
+
+func (p sqlProcessStoreImpl) doProcessLocalQueueMessageTx(
+	ctx context.Context, tx extensions.SQLTransaction, request persistence.ProcessLocalQueueMessagesRequest,
+) (*persistence.ProcessLocalQueueMessagesResponse, error) {
+	assignedStateExecutionIdToMessagesMap := map[string][]persistence.LocalQueueMessageInfoJson{}
+	finishedStateExecutionIdToPreviousVersionMap := map[string]int32{}
+
+	// Step 1: update process execution row
+	prcRow, err := tx.SelectProcessExecutionForUpdate(ctx, request.ProcessExecutionId)
+	if err != nil {
+		return nil, err
+	}
+
+	waitingQueues, err := persistence.NewStateExecutionWaitingQueuesFromBytes(prcRow.StateExecutionWaitingQueues)
+	if err != nil {
+		return nil, err
+	}
+
+	// merge waitingQueues.UnconsumedMessages into request.Messages to consume
+	// waitingQueues.UnconsumedMessages need to be first to ensure the consumption order
+	messages := append(waitingQueues.UnconsumedMessages, request.Messages...)
+	waitingQueues.ClearUnconsumedMessages()
+
+	for _, message := range messages {
+		assignedStateExecutionIdString, hasFinishedWaiting := waitingQueues.Consume(message)
```

Comment:
Nit: it may be more clear to split this into two method:
* ProcessMessage(which put into the unconsumed list)
* TryConsume(see if any messages can be consumed after adding a new message)